src/borg/crypto/file_integrity.py

```diff
@@ -0,0 +1,183 @@
+        self.hash.update(str(self.tell()).encode())
+
+
+class FileIntegrityError(IntegrityError):
+    """File failed integrity check: {}"""
+
+
+class IntegrityCheckedFile(FileLikeWrapper):
+    def __init__(self, path, write, filename=None, override_fd=None):
+        self.path = path
+        self.writing = write
+        mode = 'wb' if write else 'rb'
+        self.file_fd = override_fd or open(path, mode)
+
+        self.fd = self.hasher = SHA512FileHashingWrapper(backing_fd=self.file_fd, write=write)
+
+        self.hash_filename(filename)
+
+        if write:
+            self.digests = {}
+        else:
+            self.digests = self.read_integrity_file(path, self.hasher)
+            # TODO: When we're reading but don't have any digests, i.e. no integrity file existed,
+            # TODO: then we could just short-circuit.
+
+    def hash_filename(self, filename=None):
+        # Hash the name of the file, but only the basename, ie. not the path.
+        # In Borg the name itself encodes the context (eg. index.N, cache, files),
+        # while the path doesn't matter, and moving e.g. a repository or cache directory is supported.
+        # Changing the name however imbues a change of context that is not permissible.
+        filename = os.path.basename(filename or self.path)
+        self.hasher.update(('%10d' % len(filename)).encode())
+        self.hasher.update(filename.encode())
+
+    @staticmethod
+    def integrity_file_path(path):
+        return path + '.integrity'
+
+    @classmethod
+    def read_integrity_file(cls, path, hasher):
+        try:
+            with open(cls.integrity_file_path(path), 'r') as fd:
+                integrity_file = json.load(fd)
+                # Provisions for agility now, implementation later, but make sure the on-disk joint is oiled.
+                algorithm = integrity_file['algorithm']
+                if algorithm != hasher.ALGORITHM:
+                    logger.warning('Cannot verify integrity of %s: Unknown algorithm %r', path, algorithm)
+                    return
+                digests = integrity_file['digests']
+                # Require at least presence of the final digest
+                digests['final']
+                return digests
+        except FileNotFoundError:
+            logger.info('No integrity file found for %s', path)
+        except (OSError, ValueError, TypeError, KeyError) as e:
+            logger.warning('Could not read integrity file for %s: %s', path, e)
+            raise FileIntegrityError(path)
+
+    def hash_part(self, partname, is_final=False):
+        if not self.writing and not self.digests:
+            return
+        self.hasher.update(partname.encode())
+        self.hasher.hash_length(seek_to_end=is_final)
+        digest = bin_to_hex(self.hasher.digest())
+        if self.writing:
+            self.digests[partname] = digest
+        elif self.digests and not compare_digest(self.digests.get(partname, ''), digest):
```

Comment:
is self.digests term needed? if it is an empty dict, the compare would fail anyway.